### ISSUE TITLE
Unable to load modules created with Bridge.Net

### ISSUE THREAD

**Issue by mikemherron** [phpv8/v8js#339]
I've been experimenting with using a custom module loader to include Common JS modules output by [Bridge.Net](https://bridge.net/) 

The first problem is that the core bridge.js file didn't seem to be able to "see" `module.exports`, so the following if statement in the file failed, meaning the Bridge object was not exported:

```
} else if (typeof module !== "undefined" && module.exports) {
 // Node
 module.exports = Bridge;
}
```

I managed to "fix" this by removing the `if` test and always setting module.exports to Bridge. This is weird, as it seems to suggest that `module` must have been available within the scope, but not `module.exports`?

Second problem was that due to (I think) the lack of the global object, the Bridge class itself was not exposed to the generated code after loading. So the generated JS code:

```
Bridge.assembly("BridgeTest", function ($asm, globals) {
    "use strict";

    (function () {
        var CS2 = { };
        Bridge.define("BridgeTest.Product", {
            $scope: CS2,
            fields: {
                Id: 0,
                Name: null
            },
            methods: {
                GetString: function () {
                    return this.Id + ":" + (this.Name || "");
                }
            }
        });
        module.exports.CS2 = CS2;
    }) ();
});
```

Would fail to execute because `Bridge.assembly` was not available. I got round this by pre-pending the below line in my module loader when loading the source:

`var Bridge = require('bridge');`

Which moved things along, but then hit another snag when the export line was executed:

`CS2:23: TypeError: Cannot set property 'CS2' of undefined`

Again, this seems strange as it looks like `module` was visible, but not `exports`. I pre-pended another line of code at the start of the generated JS:

`module.exports = {};`

This allowed the module to be included (finally!). At that point I was able to create an instance of the Product class, and set properties:

```
var testModule = require("CS2");
var product = new testModule.CS2.BridgeTest.Product();
product.Name = "A produt";
product.Id = 124;
```

But attempting to call the GetString method failed: 

`TypeError: product.GetString is not a function`

Iterating through the properties of `product` in the above code suggests that Bridge.Net is doing some sort of processing and creating the Product definition:

```
"$init"
"Name"
"Id"
"constructor"
"$$fullname"
"$scope"
"fields"
"methods"
"$kind"
"ctor"
"$initialize"
"$$name"
"toJSON"
```

But for whatever reason, the `GetString` method is not working correctly. The code all works when included in a standalone HTML file. 

So in summary, it doesn't feel like it's worth looking any deeper in to the final method error as it seems like there are fundamental differences in the way modules are loaded between php v8js and Node.JS which might be causing even more issues that I just haven't discovered yet.

I know that the lack of the global object has already been documented (#267) but the module.exports problem seems really odd. Any suggestions on potential fixes or workarounds would be appreciated. I know @stesie provided a reference to a commit that included the global object in #267, but not sure how up to date this is.

**Comment by mikemherron** [comment]
So manually compiling with the commit referenced in #267 (https://github.com/stesie/v8js/commit/1ae99d2ab945b0900a81c106a26ea83b18df0d11) hacked in to master solves the global problem, and also means that Bridge defined methods now work as shown with the below test script:

```
<?php

$JS = <<< EOT
require("bridge");
var module = require("somemodule");

var product = new module.CS2.BridgetTest.Product();
product.Id = 1;
product.Name = "Test";
var_dump(product.GetString());
EOT;

$v8 = new V8Js();
$v8->setModuleLoader(function($module) {
  if ($module=="bridge") {
    	return file_get_contents("https://raw.githubusercontent.com/bridgedotnet/Bridge/master/Bridge/Resources/.generated/bridge.js");
	//return file_get_contents(__DIR__ . "/bridge.js");
  }
  else {
    return '
    //manual initialisation of module.exports still required..
    module.exports = {};

    Bridge.assembly("BridgetTest", function ($asm, globals) {
        "use strict";

        (function () {
            var CS2 = { };
            Bridge.define("BridgetTest.Product", {
                $scope: CS2,
                fields: {
                    Id: 0,
                    Name: null
                },
                methods: {
                    GetString: function () {
                        return this.Id + ":" + (this.Name || "");
                    }
                }
            });
            module.exports.CS2 = CS2;
        }) ();
    });
';
  }
});

$v8->executeString($JS);
?>
```

The module.exports object must still be manually defined, for my use case that's probably something I can work around. 

Is there any chance of global support being included in master anytime soon? I'm happy to contribute with a PR, but wanted to check there are no reasons it would be rejected straight off. I'm guessing backwards compatibility may be an issue?

**Comment by stesie** [comment]
Hi @mikemherron,

sorry for not answering earlier, and I very much appreciate that you kept investigating. Thank you for that! And good to know that it'd help to make v8js more node-compatible.

Unfortunately I'm currently a bit short on time and generally have no more a company backing my effort towards v8js. And beyond that require() stuff never was high on my personal priority list, I mainly focused on data passing.

Anyways I'm personally all for making v8js more compatible towards node and wouldn't refrain from merging such stuff in. I'd consider it more of a bug (i.e. it currently misbehaves) and not a new (breaking) feature

That said the branch you've mentioned above isn't yet in good shape, I've started refactoring the require-code once, yet then got lost halfway through and the patch causes some memory issues. So I'd be careful to merge all of it yet. Adapting (cherry-picking) just the linked commit itself seems sensible to me though (and actually we really should refactor this looooongish function sooner than later).

TL;DR if you'd like to try doing so, let me know and go for it. I'm more then willing to merge it, once well in shape. Also I might have an hour or two on the weekend to maybe look myself.

Regarding `module.exports` I've just had a brief look into the source, it currently just provides `exports = {}` before evaluating the module and then picks up either `modules.exports` (if created) or otherwise `exports`. 

A current node version provides `exports = module.exports = {}`, so I think it'd make sense to mimic at least that. Apart from that node always returns `module.exports` as the "result", so we might even want to always return just `module.exports` as well (and not fall back to `modules`).
I'm currently unsure whether to do also the latter, yet I tend to it

**Comment by mikemherron** [comment]
Thanks for getting back to me @stesie - that sounds good, I will have more free time in the next 2 weeks so will look at putting together a proper PR based on your commit. Feel free to beat me to it though if you manage to find time!! Thanks again for your feedback and all your work on v8js.

**Comment by stesie** [comment]
Hey, just wanted to let you know that I've prepared a pull request #340 with both things you wanted to have: i.e. `global` and `module.exports` being available and behaving like on Node.js
Please try it out and let me know what you think :-)

If it works I'll merge it and then probably release a version 2.0.0; i.e. bump major number, just to make the change clear. Actually I wanted to release 2.0.0 a little later (and then remove deprecated stuff with it), but now it's more likely that there will be a version 3.0.0 sooner :)

**Comment by mikemherron** [comment]
Brilliant, that's great! I'll try this today and report back.

**Comment by mikemherron** [comment]
I've built against your branch @stesie and re-ran my test - everything works great. The Bridge object is now available globally after requiring the module and explicit declaration of `module.exports` is no longer required! Awesome, great stuff.

**Comment by mikemherron** [comment]
So, interestingly, I've tried moving the built library and v8 libs from my build environment to a clean development environment and am now getting other interesting errors that I hadn't seen before regarding the `$asm` and `$scope` variables being undefined - but not the `Bridge` variable oddly. It's as if global scope is now working (or `Bridge` itself would not be defined) but something else is not correct. I'm 99% sure this is an environment issue on my side though; I'm going to keep investigating.

**Comment by mikemherron** [comment]
Ignore me....I'm an idiot. Updated v8js is totally fine (as expected), was using an old, broken version of my test script that declared the JS string using heredoc, causing the $-prefixed variables to be escaped in to PHP. The error was a PHP error that the variables are not found, nothing to do with v8JS, I just wasn't reading it properly and jumping to conclusions.

TL;DR - everything looks good with the PR!
